### PR TITLE
Ft shapes 

### PR DESCRIPTION
**Pull request by Roxana-P**
The feature implements a collection of geometric primitives as splines as solution to issue #46 . 
The geometries are implemented in `gustaf.spline.create`. They are created by e.g.
`bezier = gustaf.spline.box(1.0, 2.0)`

The routines return the simplest spline type possible with a minimum number of knot elements and minimum degree. 
Most circular objects are created by the revolve method.

### CODE REVIEW

File: gustaf/spline/create.py [hunk 1]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+
+    cps = np.array(
+            [
+                    [r, 0., 0.], [r, -r, 0.], [0., -r, 0.], [-r, -r, 0.],
+                    [-r, 0., 0.], [-r, r, 0.], [0., r, 0.], [r, r, 0.],
+                    [r, 0., 0.]
+            ]
+    )
+
+    weights = np.tile([1.0, 1 / np.sqrt(2)], 5)[:-1]
```
Comment (jzwar) [inline gustaf/spline/create.py:314]: Clever :wink:

File: gustaf/spline/create.py [hunk 2]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+                    np.full(degree,
+                            0.), np.linspace(0., 1., nknots - 2 * degree),
+                    np.full(degree, 1.0)
+            )
+    )
+
+    return BSpline(control_points=cps, knot_vectors=[knots], degrees=[degree])
+
+
+def circle(r):
```
Comment (jzwar) [inline gustaf/spline/create.py:289]:
```suggestion
def circle(radius):
```

File: gustaf/spline/create.py [hunk 3]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+
+    return NURBS(
+            control_points=cps,
+            knot_vectors=[knots],
+            degrees=[degree],
+            weights=weights
+    )
+
+
+def rectangle(a, b):
```
Comment (jzwar) [inline gustaf/spline/create.py:326]:
```suggestion
def rectangle(width, length):
```

File: gustaf/spline/create.py [hunk 4]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+
+    cps = np.zeros([4, 3])
+    cps[1, 0] = a
+    cps[2, 1] = b
+    cps[3, :2] = a, b
+
+    return Bezier(control_points=cps, degrees=[1, 1])
+
+
+def box(a, b, h):
```
Comment (jzwar) [inline gustaf/spline/create.py:351]:
```suggestion
def box(width, length, height):
```

File: gustaf/spline/create.py [hunk 5]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+      dimension of the box in y-direction
+    h : float
+      dimension of the box in z-direction
+
+    Returns
+    -------
+    gustaf.Bezier
+      Volumetric linear Bezier spline
+    """
+    return rectangle(a, b).create.extruded([0., 0., h])
```
Comment (jzwar) [inline gustaf/spline/create.py:369]: Nice!

File: gustaf/spline/create.py [hunk 6]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+
+
+def disk(R, r0=0., angle=360., n_knot_spans=4):
```
Comment (jzwar) [inline gustaf/spline/create.py:372]:
```suggestion
def disk(outer_radius, inner_radius=0., angle=360., n_knot_spans=4):
```
Try to be a bit more describtive with your arguments

File: gustaf/spline/create.py [hunk 7]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+    cps = np.array([[r0, 0., 0.], [R, 0., 0.]])
+    weights = np.ones([cps.shape[0]])
+    knots = np.repeat([0., 1.], 2)
+
+    return NURBS(
+            control_points=cps,
+            knot_vectors=[knots],
+            degrees=[1],
+            weights=weights
+    ).create.revolved(
```
Comment (jzwar) [inline gustaf/spline/create.py:405]: Why do you perform the rotation in 3D, wouldn't 2D be sufficient here?

File: gustaf/spline/create.py [hunk 8]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+            axis=[0., 0., 1.],
+            center=np.zeros(3),
+            angle=angle,
+            n_knot_spans=n_knot_spans
+    )
+
+
+def torus(R, r, r0=0., angle=[360., 360.], n_knot_spans=[4, 4]):
```
Comment (jzwar) [inline gustaf/spline/create.py:413]: see above

File: gustaf/spline/create.py [hunk 9]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+      direction, by default both equal 4
+
+    Returns
+    -------
+    gustaf.NURBS
+      Volumetric spline in the shape of a torus with degrees (1,2,2)
+    """
+
+    # Create a surface spline representing a disk and move it from the origin
+    d = disk(r, r0, n_knot_spans=n_knot_spans[1], angle=angle[1])
```
Comment (jzwar) [inline gustaf/spline/create.py:438]: Computationally I would prefer to have a tube-torus defined differrently from a full torrus(-segment). I.e., have a spline like this:
```
plate = gus.RationalBezier(
        degrees=[2,2],
        control_points=[
                                [0, 0], 
                                [0.5, -0.5],
                                [1, 0],
                                [-0.5, 0.5], 
                                [0.5, 0.5], 
                                [1 + 0.5, 0.5],
                                [0, 1],
                                [0.5, 1 + 0.5],
                                [1, 1],
                        ],
        weights=[1, isqrt2, 1, isqrt2, 1, isqrt2, 1, isqrt2, 1]
)
```
Be revolved to a torus

**Review by jzwar — COMMENTED** [review]
Hey, 
thanks for these helper functions, they really make it much easier to define splines in gustaf.
Although they look correct on first sight, they are not quite in the style of the rest of gustaf. I made a couple of suggestions and would ask you to update the code accordingly.
Thereafter you can re-request a review and I will have a look at the rest

File: gustaf/spline/create.py [hunk 10]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+    degree : int, optional
+      Desired spline degree, by default 1
+
+    Returns
+    -------
+    gustaf.Bspline
+      Bspline with choosen control points and if necessary internal knots.
+    """
+    from gustaf import BSpline
+    cps = points
```
[the remark this answers is not in the document]
Comment (Roxana-P) [reply gustaf/spline/create.py:275]: I changed the line to 
`cps = np.array(points)`
which fails if points is not an array / list-like. Is that sufficient or do you expect an explicit exception in that case?

[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/create.py:275]: that's fine with me

File: gustaf/spline/create.py [hunk 11]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+    nknots = cps.shape[0] + degree + 1
+
+    knots = np.concatenate(
+            (
+                    np.full(degree,
+                            0.), np.linspace(0., 1., nknots - 2 * degree),
+                    np.full(degree, 1.0)
```
[the remark this answers is not in the document]
Comment (Roxana-P) [reply gustaf/spline/create.py:282]: Ok

File: gustaf/spline/create.py [hunk 12]
```diff
@@ -254,6 +255,289 @@ def revolved(
...
+      Desired spline degree, by default 1
+
+    Returns
+    -------
+    gustaf.Bspline
+      Bspline with choosen control points and if necessary internal knots.
+    """
+    from gustaf import BSpline
+    cps = points
+    nknots = cps.shape[0] + degree + 1
```
[the remark this answers is not in the document]
Comment (Roxana-P) [reply gustaf/spline/create.py:276]: Ok

File: gustaf/spline/create.py [hunk 13]
```diff
@@ -254,6 +255,289 @@ def revolved(
     return type(spline)(**spline_dict)
 
 
+def line(points, degree=1):
```
[the remark this answers is not in the document]
Comment (Roxana-P) [reply gustaf/spline/create.py:258]: No, weights are only useful for circular arcs, which can be created with the newly-added routine arc()

File: gustaf/spline/create.py (hunk 6 above)
Comment (Roxana-P) [reply gustaf/spline/create.py:372]: ****

File: gustaf/spline/create.py [hunk 14]
```diff
@@ -254,6 +255,397 @@ def revolved(
...
+    """
+    if torus_angle is None:
+        torus_angle = 360
+    if section_angle is None:
+        section_angle = 360
+    if section_inner_radius is None:
+        section_inner_radius = 0
+
+    # Create the cross-section
+    if section_angle == 360 and section_inner_radius == 0:
```
Comment (jzwar) [inline gustaf/spline/create.py:523]: Float Comparisons are always kind of dangerous, maybe put an additional boolean if angle is None

File: gustaf/spline/create.py [hunk 15]
```diff
@@ -254,6 +255,397 @@ def revolved(
...
+      Number of knot spans
+
+    Returns
+    -------
+    gustaf.NURBS
+      Volumetric NURBS with degrees (1,2,2)
+    """
+
+    if inner_radius is None:
+        sphere = plate(outer_radius).nurbs.create.revolved(
```
Comment (jzwar) [inline gustaf/spline/create.py:569]: This will also produce a warning if i am correct

File: gustaf/spline/create.py [hunk 16]
```diff
@@ -254,6 +255,397 @@ def revolved(
...
+        sphere = disk(outer_radius, inner_radius).nurbs.create.revolved(
+                axis=[1, 0, 0],
+                center=[0, 0, 0],
+                angle=angle,
+                n_knot_spans=n_knot_spans
+        )
+    return sphere
+
+
+def cone(radius, height, volumetric=True, angle=360.):
```
Comment (jzwar) [inline gustaf/spline/create.py:586]: Please let the user define two radii to make it more accessible

File: gustaf/spline/create.py [hunk 17]
```diff
@@ -254,6 +255,397 @@ def revolved(
...
+      Volumetric or surface NURBS descibing a cone
+    """
+
+    if volumetric:
+        ground = disk(radius, angle=angle)
+    else:
+        ground = circle(radius)
+
+    # Extrude in z
+    con = ground.create.extruded([0, 0, height])
```
Comment (jzwar) [inline gustaf/spline/create.py:612]:
```suggestion
    cone_ = ground.create.extruded([0, 0, height])
```

File: gustaf/spline/create.py [hunk 18]
```diff
@@ -254,6 +255,397 @@ def revolved(
...
+    gustaf.Bspline
+      Volumetric linear spline in the shape of a pyramid
+    """
+
+    # Create box
+    p = box(a, b, h)
+
+    # Collapse all upper points on one control point
+    p.control_points[np.isclose(p.control_points[:, -1],
+                                h)] = [a / 2, b / 2, h]
```
Comment (jzwar) [inline gustaf/spline/create.py:644]:
```suggestion
    p.control_points[:, 4:] = [a / 2, b / 2, h]
```

**Review by jzwar — COMMENTED** [review]
Hey, great improvement, thanks! There is still a few things that need refinement.

**Comment by jzwar** [comment]
This branch needs rebasing there are now merge conflicts.

**Comment by clemens-fricke** [comment]
I am currently getting the following error when I am trying to run the show_geometries.py file. 

The environment has only splinepy (latest repo version)  and gustaf(this pr) installed.

image

File: gustaf/spline/create.py [hunk 19]
```diff
@@ -254,6 +255,437 @@ def revolved(
...
+    if section_inner_radius is None:
+        section_inner_radius = 0
+        section_inner_radius_flag = False
+
+    # Create the cross-section
+    if section_angle_flag is False and section_inner_radius_flag is False:
+        cross_section = plate(section_outer_radius)
+        # For more than 180 degree only NURBS can be used
+        if torus_angle >= 180:
+            cross_section = cross_section.nurbs
```
Comment (j042) [inline gustaf/spline/create.py:548]:
```suggestion
    if torus_angle is None:
        torus_angle = 2 * np.pi
    if section_angle is None:
        section_angle = 2 * np.pi
        section_angle_flag = False
    if section_inner_radius is None:
        section_inner_radius = 0
        section_inner_radius_flag = False

    # Create the cross-section
    if not section_angle_flag and not section_inner_radius_flag:
        cross_section = plate(section_outer_radius)
        # For more than 180 degree only NURBS can be used
        if abs(torus_angle % (2 * np.pi)) >= np.pi:
            cross_section = cross_section.nurbs
```

**Review by j042 — COMMENTED** [review]
Seems very useful, thank you!  
What about we name the example, `show_spline_geometries` to highlight spline?

File: gustaf/spline/create.py [hunk 20]
```diff
@@ -254,6 +255,437 @@ def revolved(
...
+
+    # Define point spline of degree 0 at starting point of the arc
+    if degree is True:
+        start_angle = np.radians(start_angle)
+    start_point = [radius * np.cos(start_angle), radius * np.sin(start_angle)]
+    point_spline = RationalBezier(
+            degrees=[0], control_points=[start_point], weights=[1.]
+    )
+    # Bezier splines only support angles lower 180 degrees
+    if angle >= 180 or n_knot_spans > 1:
```
[the remark this answers is not in the document]
Comment (Roxana-P) [reply gustaf/spline/create.py]: I adjusted the use of radians. The use of the modulus does not have the intended effect here, as, e.g., angles of 180 degrees and 360 need a different number of knot spans. 
The input of degrees > 360 leads to physically overlapping, but mathematically valid splines, which can be visualized.

**Comment by Roxana-P** [comment]
> I am currently getting the following error when I am trying to run the show_geometries.py file.
> 
> The environment has only splinepy (latest repo version) and gustaf(this pr) installed.
> 
> image
Problem is resolved

File: gustaf/spline/create.py [hunk 21]
```diff
@@ -254,6 +255,448 @@ def revolved(
...
+
+    Returns
+    -------
+    gustaf.NURBS or gustaf.RationalBezier
+        Line spline describing an arc
+    """
+    from gustaf import RationalBezier
+
+    # Define point spline of degree 0 at starting point of the arc
+    if degree is True:
```
Comment (j042) [inline gustaf/spline/create.py:323]:
```suggestion
    if degree:
```

**Review by j042 — APPROVED** [review]
Awesome, thank you!

**Comment by j042** [comment]
I think your branch isn't properly rebased @Roxana-P, could you please double check?

**Comment by j042** [comment]
error message from ci
```
File "/home/runner/work/gustaf/gustaf/gustaf/spline/create.py", line 332
    <<<<<<< HEAD
    ^
SyntaxError: invalid syntax
```

**Review by j042 — DISMISSED** [review]

**Review by j042 — APPROVED** [review]
Thank you @Roxana-P!

**Pull request merged**